pyfr/plugins/ascent.py

```diff
@@ -0,0 +1,425 @@
+        cons = self.cfg.items_as('constants', float)
+        scene_cfg = {}
+        self.scenes = ConduitNode(self.conduit)
+
+        for k in self.cfg.items(cfgsect):
+            if k.startswith('scene-'):
+                configs = self.cfg.getliteral(cfgsect, k)
+                field = configs['field']
+                sn = k.removeprefix('scene-')
+
+                scene_cfg[sn] = configs
+
+                for kc, vc in configs.items():
+                    if kc == 'image_name':
+                        pass
+                    elif kc == 'field_expr':
+                        if field in self._expr:
+                            raise KeyError(f"Field '{field}' already exists")
+                        self._expr[field] = self._get_expr(vc, subs=cons)
+                    elif kc.startswith('field_expr_'):
+                        sub_field = kc.removeprefix('field_expr_')
+                        if sub_field not in 'xyz':
+                            raise KeyError('Invalid vector subfield ID')
+                        if field not in self._expr:
+                            self._expr[field] = {}
+                        elif not isinstance(self._expr[field], dict):
+                            raise KeyError(f"Field  '{field}' already exists")
+                        expr = self._get_expr(vc, subs=cons)
+                        self._expr[field][sub_field] = expr
+                    elif kc == 'pipeline':
+                        self.scenes[f's_{sn}/plots/p1/pipeline'] = f'pl_{vc}'
+                    elif kc == 'field':
+                        self.scenes[f's_{sn}/plots/p1/{kc}'] = field
+                    elif kc.startswith('render-'):
+                        ropt = kc.removeprefix('render-')
+                        self.scenes[f's_{sn}/renders/r1/{ropt}'] = vc
+                    else:
+                        self.scenes[f's_{sn}/plots/p1/{kc}'] = vc
+        self._scene_cfg = scene_cfg
+
+    def _config_pipeline(self, cfgsect):
+        pipel_cfg = {}
+        self.pipelines = pl = ConduitNode(self.conduit)
+
+        for k in self.cfg.items(cfgsect):
+            if k.startswith('pipeline-'):
+                configs = self.cfg.getliteral(cfgsect, k)
+                configs = [configs] if isinstance(configs, dict) else configs
+                pn = k.removeprefix('pipeline-')
+
+                pipel_cfg[pn] = configs
+                for j, filter in enumerate(configs):
+                    params = ConduitNode(self.conduit)
+                    for kf, vf in filter.items():
+                        if kf == 'type':
+                            ptype = vf
+                            pl[f'pl_{pn}/f{j}/{kf}'] = vf
+                        elif kf == 'output_name':
+                            if vf in  self._expr:
+                                raise KeyError(f"Output name '{vf}' already used")
+                            params[kf] = vf
+                            self._expr[vf] = None
```

Comment:
I'd be inclined to use a pair of variables for this: a set to track definitions (empty and non-empty) and then just have exprs contain actual populated expressions.